src/wccp.cc

```diff
@@ -167,6 +168,29 @@ wccpConnectionClose(void)
     }
 }
 
+class WccpRr : public RegisteredRunner
+{
+public:
+    virtual void useConfig() override {
+        if (IamPrimaryProcess()) {
+            wccpInit();
+            wccpConnectionOpen();
+        }
+    }
+
+    // reconfigure fully shuts down, then restarts WCCP
+    virtual void startReconfigure() override { startShutdown(); }
+    virtual void syncConfig() override { useConfig(); }
+
+    // must run before comm_exit()
```

Comment:
The RunnersRegistry documentation does not mention comm in any way. It mentions only the other Runner methods timing and a "main loop" (whatever that is. Yes **we** two know, but other dev may not).

This comment is documenting why the method override is done on startShutdown() instead of finishShutdown().

Reply:
```C++
   /// Called after receiving a shutdown request and before stopping the main
    /// loop. At least one main loop iteration is guaranteed after this call.
    /// Meant for cleanup and state saving that may require other modules.
    virtual void startShutdown() {}
```

> Alex: The existing startShutdown() event description already strongly implies that the event is triggered before comm_exit()

> Amos: The RunnersRegistry documentation does not mention comm in any way.

IMO, that documentation does not have to mention "comm" explicitly to strongly imply that startShutdown() event is triggered before comm_exit()! Guarantying at least one main loop iteration is enough because the general purpose main loop (used by lots of Comm-dependent code) should not run without such a fundamental piece as the Comm module providing its services. Moreover, _all_ lower-level modules/services should be available during main loop iterations! We do not want to constantly add `if Comm is still there` and similar guards to regular code that gets triggered during main loop iterations.

Any event hook (i.e. an override of the documented virtual RegisteredRunner method) can only rely on the documented[^1] event timing/preconditions. The current PR comment looks very much like a new (misplaced or unnecessary) precondition on the event. Hence this change request.

The specific suggestion in this change request is based on my opinion that (what looks like a precondition) documentation added by this PR is unnecessary. If that opinion is proven to be wrong, then we should modify the RegisteredRunner documentation (and still remove this specific comment!) to address this change request.

[^1]: If the startReconfigure() documentation is defective, we should improve it, of course.
 
> This comment is documenting why the method override is done on startShutdown() instead of finishShutdown().

The comment does not achieve the intended effect and, again, is strangely focusing on one out of several existing dependencies. If you insist, you can modify this comment to remove those problems. For example:

```suggestion
    // we use startShutdown() instead of later events because 
    // Comm socket closures require main loop iterations
```